File: codes/matlab/ch03/collocation_vs_galerkin.m
%% collocation_vs_galerkin.m
%
% Compares the Collocation (pseudospectral) and Galerkin methods for solving
% a reaction-diffusion boundary value problem.
%
% Problem:
%     u''(x) - 4u(x) = -1,   -1 ≤ x ≤ 1
%     u(-1) = 0,  u(1) = 0
%
% Exact solution:
%     u_exact(x) = (1/4)(1 - cosh(2x)/cosh(2))
%
% Basis functions (satisfy homogeneous BCs):
%     φ₀(x) = (1 - x²)
%     φ₁(x) = (1 - x²)x² = x² - x⁴
%
% Trial function:
%     u₁(x) = a₀φ₀(x) + a₁φ₁(x)
%
% Methods compared:
%     1. Collocation: Force residual to zero at x = 0 and x = 0.5
%     2. Galerkin: Force residual orthogonal to basis functions
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Études: A Spectral Approach"

clear; close all; clc;

%% Configuration
N_POINTS = 500;

% Book color scheme
NAVY = [20, 45, 110] / 255;
SKY = [120, 150, 210] / 255;
CORAL = [231, 76, 60] / 255;
GREEN = [39, 174, 96] / 255;

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch03', 'matlab');
output_file = fullfile(output_dir, 'collocation_vs_galerkin.pdf');

%% Define functions
% Exact solution
u_exact = @(x) 0.25 * (1 - cosh(2*x) / cosh(2));

% Basis functions
phi0 = @(x) 1 - x.^2;
phi1 = @(x) x.^2 - x.^4;

% Second derivatives
phi0_xx = @(x) -2 * ones(size(x));
phi1_xx = @(x) 2 - 12*x.^2;

% Operator L = d²/dx² - 4
L_phi0 = @(x) phi0_xx(x) - 4*phi0(x);
L_phi1 = @(x) phi1_xx(x) - 4*phi1(x);

%% Solve Collocation system
% Collocation points: x = 0 and x = 0.5
x1 = 0.0;
x2 = 0.5;

% Build matrix A where A*[a0; a1] = b
A_coll = [L_phi0(x1), L_phi1(x1);
          L_phi0(x2), L_phi1(x2)];

% RHS: f = -1, so R = Lu - f = Lu + 1 = 0 means Lu = -1
b_coll = [-1; -1];

% Solve
coeffs_coll = A_coll \ b_coll;
a0_coll = coeffs_coll(1);
a1_coll = coeffs_coll(2);

fprintf('Collocation coefficients:\n');
fprintf('  a₀ = %.6f\n', a0_coll);
fprintf('  a₁ = %.6f\n', a1_coll);

%% Solve Galerkin system
% Matrix entries: A_{ij} = ∫_{-1}^{1} L[φⱼ](x) φᵢ(x) dx

A00 = integral(@(x) L_phi0(x) .* phi0(x), -1, 1);
A01 = integral(@(x) L_phi1(x) .* phi0(x), -1, 1);
A10 = integral(@(x) L_phi0(x) .* phi1(x), -1, 1);
A11 = integral(@(x) L_phi1(x) .* phi1(x), -1, 1);

A_gal = [A00, A01; A10, A11];

% RHS: b_i = ∫_{-1}^{1} f(x) φᵢ(x) dx = ∫_{-1}^{1} (-1) φᵢ(x) dx
b0 = integral(@(x) -1 * phi0(x), -1, 1);
b1 = integral(@(x) -1 * phi1(x), -1, 1);

b_gal = [b0; b1];

% Solve
coeffs_gal = A_gal \ b_gal;
a0_gal = coeffs_gal(1);
a1_gal = coeffs_gal(2);

fprintf('\nGalerkin coefficients:\n');
fprintf('  a₀ = %.6f\n', a0_gal);
fprintf('  a₁ = %.6f\n', a1_gal);

%% Define approximate solutions
u_coll_fn = @(x) a0_coll * phi0(x) + a1_coll * phi1(x);
u_gal_fn = @(x) a0_gal * phi0(x) + a1_gal * phi1(x);

%% Compute solutions on grid
x = linspace(-1, 1, N_POINTS)';
u_ex = u_exact(x);
u_coll_val = u_coll_fn(x);
u_gal_val = u_gal_fn(x);

error_coll = u_ex - u_coll_val;
error_gal = u_ex - u_gal_val;

% Collocation points for visualization
x_coll_pts = [0.0; 0.5];
u_coll_pts = u_coll_fn(x_coll_pts);

%% Create two-panel figure
fig = figure('Units', 'inches', 'Position', [1, 1, 9, 3.5]);

% Left panel: All solutions
subplot(1, 2, 1);
hold on; box on;

plot(x, u_ex, 'Color', NAVY, 'LineWidth', 1.8, 'DisplayName', 'Exact');
plot(x, u_coll_val, '--', 'Color', CORAL, 'LineWidth', 1.5, ...
     'DisplayName', 'Collocation');
plot(x, u_gal_val, '-.', 'Color', GREEN, 'LineWidth', 1.5, ...
     'DisplayName', 'Galerkin');
plot(x_coll_pts, u_coll_pts, 's', 'Color', CORAL, 'MarkerSize', 8, ...
     'MarkerFaceColor', 'none', 'LineWidth', 1.5, ...
     'DisplayName', 'Collocation pts');

xlabel('$x$', 'Interpreter', 'latex');
ylabel('$u(x)$', 'Interpreter', 'latex');
xlim([-1, 1]);
title('Solutions Comparison', 'FontSize', 11);
legend('Location', 'south', 'Interpreter', 'latex', 'NumColumns', 2);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

% Right panel: Errors
subplot(1, 2, 2);
hold on; box on;

plot(x, error_coll, 'Color', CORAL, 'LineWidth', 1.5, 'DisplayName', 'Collocation');
plot(x, error_gal, 'Color', GREEN, 'LineWidth', 1.5, 'DisplayName', 'Galerkin');
yline(0, '--', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5);

xlabel('$x$', 'Interpreter', 'latex');
ylabel('$u_{\mathrm{exact}} - u_{\mathrm{approx}}$', 'Interpreter', 'latex');
xlim([-1, 1]);
title('Error Comparison', 'FontSize', 11);
legend('Location', 'northeast', 'Interpreter', 'latex');

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

%% Save figure
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

exportgraphics(fig, output_file, 'ContentType', 'vector');
fprintf('\nFigure saved to: %s\n', output_file);

png_file = strrep(output_file, '.pdf', '.png');
exportgraphics(fig, png_file, 'Resolution', 300);

%% Print comparison table
fprintf('\n%s\n', repmat('=', 1, 65));
fprintf('Comparison at u(0) - the central maximum\n');
fprintf('%s\n', repmat('=', 1, 65));
fprintf('%-20s %-15s %-15s\n', 'Method', 'u(0)', 'Abs. Error');
fprintf('%s\n', repmat('-', 1, 65));
fprintf('%-20s %-15.6f %-15.6f\n', 'Exact', u_exact(0), 0);
fprintf('%-20s %-15.6f %-15.6f\n', 'Collocation (N=2)', ...
        u_coll_fn(0), abs(u_exact(0) - u_coll_fn(0)));
fprintf('%-20s %-15.6f %-15.6f\n', 'Galerkin (N=2)', ...
        u_gal_fn(0), abs(u_exact(0) - u_gal_fn(0)));
fprintf('%s\n', repmat('=', 1, 65));

% RMS errors
rms_coll = sqrt(mean(error_coll.^2));
rms_gal = sqrt(mean(error_gal.^2));
fprintf('\nRMS Error (Collocation): %.6f\n', rms_coll);
fprintf('RMS Error (Galerkin):    %.6f\n', rms_gal);

% Max errors
max_coll = max(abs(error_coll));
max_gal = max(abs(error_gal));
fprintf('\nMax Error (Collocation): %.6f\n', max_coll);
fprintf('Max Error (Galerkin):    %.6f\n', max_gal);

File: codes/julia/ch03/collocation_vs_galerkin.jl
# collocation_vs_galerkin.jl
#
# Compares the Collocation (pseudospectral) and Galerkin methods for solving
# a reaction-diffusion boundary value problem. This example illustrates two
# fundamental approaches to the Method of Weighted Residuals.
#
# Problem:
#     u''(x) - 4u(x) = -1,   -1 <= x <= 1
#     u(-1) = 0,  u(1) = 0
#
# Exact solution:
#     u_exact(x) = (1/4)(1 - cosh(2x)/cosh(2))
#
# Basis functions (satisfy homogeneous BCs):
#     phi_0(x) = (1 - x^2)
#     phi_1(x) = (1 - x^2)*x^2
#
# Trial function:
#     u_1(x) = a_0*phi_0(x) + a_1*phi_1(x)
#
# Methods compared:
#     1. Collocation: Force residual to zero at x = 0 and x = 0.5
#     2. Galerkin: Force residual orthogonal to basis functions
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using QuadGK
using Printf

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N_POINTS = 500

# Book color scheme
const NAVY  = colorant"#142D6E"
const SKY   = colorant"#7896D2"
const CORAL = colorant"#E74C3C"
const GREEN = colorant"#27AE60"

# Output paths
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch03", "julia")

# Publication-quality theme
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Exact solution
# -----------------------------------------------------------------------------
"""
    u_exact(x)

Exact solution: u(x) = (1/4)(1 - cosh(2x)/cosh(2)).

Verification:
    u'(x)  = -(1/4)(2sinh(2x)/cosh(2)) = -(1/2)sinh(2x)/cosh(2)
    u''(x) = -cosh(2x)/cosh(2)

Substituting into ODE:
    u'' - 4u = -cosh(2x)/cosh(2) - 4*(1/4)*(1 - cosh(2x)/cosh(2))
             = -cosh(2x)/cosh(2) - 1 + cosh(2x)/cosh(2)
             = -1

Boundary conditions:
    u(+/-1) = (1/4)(1 - cosh(2)/cosh(2)) = (1/4)(1 - 1) = 0
"""
u_exact(x) = 0.25 * (1 - cosh(2x) / cosh(2))

# -----------------------------------------------------------------------------
# Basis functions
# -----------------------------------------------------------------------------
"""phi_0(x) = 1 - x^2 (vanishes at x = +/-1)"""
phi0(x) = 1 - x^2

"""phi_1(x) = x^2 - x^4 = (1 - x^2)*x^2 (vanishes at x = +/-1)"""
phi1(x) = x^2 - x^4

"""Second derivative: phi_0''(x) = -2"""
phi0_xx(x) = -2.0

"""Second derivative: phi_1''(x) = 2 - 12x^2"""
phi1_xx(x) = 2 - 12x^2

# -----------------------------------------------------------------------------
# Collocation method
# -----------------------------------------------------------------------------
"""
    solve_collocation()

Solve using collocation at x = 0 and x = 0.5.

The residual is:
    R(x) = u_1''(x) - 4*u_1(x) + 1
         = a_0*(phi_0'' - 4*phi_0) + a_1*(phi_1'' - 4*phi_1) + 1

Setting R(x) = 0 at the collocation points yields a 2x2 linear system.
"""
function solve_collocation()
    # Operator L = d^2/dx^2 - 4
    L_phi0(x) = phi0_xx(x) - 4 * phi0(x)
    L_phi1(x) = phi1_xx(x) - 4 * phi1(x)

    # Collocation points
    x1, x2 = 0.0, 0.5

    # Build system matrix
    A = [L_phi0(x1) L_phi1(x1);
         L_phi0(x2) L_phi1(x2)]

    # Right-hand side: f = -1, so R = Lu - f = Lu + 1 = 0 means Lu = -1
    b = [-1.0, -1.0]

    # Solve
    coeffs = A \ b

    return coeffs[1], coeffs[2]
end

# -----------------------------------------------------------------------------
# Galerkin method
# -----------------------------------------------------------------------------
"""
    solve_galerkin()

Solve using Galerkin method: <R, phi_k> = 0 for k = 0, 1.

This requires computing integrals:
    A_{ij} = integral_{-1}^{1} (phi_i'' - 4*phi_i) * phi_j dx
    b_i    = integral_{-1}^{1} (-1) * phi_i dx

The matrix is symmetric for self-adjoint operators.
"""
function solve_galerkin()
    # Operator L = d^2/dx^2 - 4
    L_phi0(x) = phi0_xx(x) - 4 * phi0(x)
    L_phi1(x) = phi1_xx(x) - 4 * phi1(x)

    # Compute matrix entries A_{ij} = <L[phi_j], phi_i>
    A00, _ = quadgk(x -> L_phi0(x) * phi0(x), -1, 1)
    A01, _ = quadgk(x -> L_phi1(x) * phi0(x), -1, 1)
    A10, _ = quadgk(x -> L_phi0(x) * phi1(x), -1, 1)
    A11, _ = quadgk(x -> L_phi1(x) * phi1(x), -1, 1)

    A = [A00 A01; A10 A11]

    # Compute RHS: b_i = <-1, phi_i> = -integral(phi_i dx)
    b0, _ = quadgk(x -> -1 * phi0(x), -1, 1)
    b1, _ = quadgk(x -> -1 * phi1(x), -1, 1)

    b = [b0, b1]

    # Solve
    coeffs = A \ b

    return coeffs[1], coeffs[2]
end

# -----------------------------------------------------------------------------
# Approximate solutions
# -----------------------------------------------------------------------------
"""Collocation/Galerkin approximation with given coefficients."""
u_approx(x, a0, a1) = a0 * phi0(x) + a1 * phi1(x)

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Solve both systems
    a0_coll, a1_coll = solve_collocation()
    a0_gal, a1_gal = solve_galerkin()

    println("Collocation coefficients:")
    @printf("  a_0 = %.6f\n", a0_coll)
    @printf("  a_1 = %.6f\n", a1_coll)
    println("\nGalerkin coefficients:")
    @printf("  a_0 = %.6f\n", a0_gal)
    @printf("  a_1 = %.6f\n", a1_gal)

    # Spatial grid
    x = range(-1, 1, length=N_POINTS)

    # Compute solutions
    u_ex = u_exact.(x)
    u_coll = u_approx.(x, a0_coll, a1_coll)
    u_gal  = u_approx.(x, a0_gal, a1_gal)

    error_coll = u_ex .- u_coll
    error_gal  = u_ex .- u_gal

    # Collocation points for visualization
    x_coll_pts = [0.0, 0.5]
    u_coll_pts = u_approx.(x_coll_pts, a0_coll, a1_coll)

    # Create two-panel figure
    fig = Figure(size=(720, 280))

    # --- Left panel: All solutions ---
    ax1 = Axis(fig[1, 1],
               xlabel = L"x",
               ylabel = L"u(x)",
               title = "Solutions Comparison",
               titlesize = 11)
    xlims!(ax1, -1, 1)

    # Exact solution (solid line)
    lines!(ax1, collect(x), u_ex, color=NAVY, linewidth=1.8,
           label="Exact")

    # Collocation approximation (dashed line)
    lines!(ax1, collect(x), u_coll, color=CORAL, linewidth=1.5,
           linestyle=:dash, label="Collocation")

    # Galerkin approximation (dash-dot line)
    lines!(ax1, collect(x), u_gal, color=GREEN, linewidth=1.5,
           linestyle=:dashdot, label="Galerkin")

    # Collocation points (open squares)
    scatter!(ax1, x_coll_pts, u_coll_pts, color=:transparent, markersize=10,
             marker=:rect, strokewidth=1.5, strokecolor=CORAL,
             label="Collocation pts")

    axislegend(ax1, position=:cb, framevisible=false, bgcolor=(:white, 0.9),
               nbanks=2)

    # Style: hide top/right spines, add subtle grid
    hidespines!(ax1, :t, :r)
    ax1.ygridvisible = true
    ax1.ygridstyle = :solid
    ax1.ygridcolor = (:gray, 0.2)

    # --- Right panel: Errors ---
    ax2 = Axis(fig[1, 2],
               xlabel = L"x",
               ylabel = L"u_{\mathrm{exact}} - u_{\mathrm{approx}}",
               title = "Error Comparison",
               titlesize = 11)
    xlims!(ax2, -1, 1)

    lines!(ax2, collect(x), error_coll, color=CORAL, linewidth=1.5,
           label="Collocation")
    lines!(ax2, collect(x), error_gal, color=GREEN, linewidth=1.5,
           label="Galerkin")
    hlines!(ax2, [0.0], color=:gray, linewidth=0.5, linestyle=:dash)

    axislegend(ax2, position=:rt, framevisible=false, bgcolor=(:white, 0.9))

    # Style: hide top/right spines, add subtle grid
    hidespines!(ax2, :t, :r)
    ax2.ygridvisible = true
    ax2.ygridstyle = :solid
    ax2.ygridcolor = (:gray, 0.2)

    # Save figure
    mkpath(OUTPUT_DIR)
    save(joinpath(OUTPUT_DIR, "collocation_vs_galerkin.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "collocation_vs_galerkin.png"), fig, px_per_unit=4)

    println("\nFigure saved to: $(joinpath(OUTPUT_DIR, "collocation_vs_galerkin.pdf"))")

    # Print comparison table
    println("\n" * "=" ^ 65)
    println("Comparison at u(0) - the central maximum")
    println("=" ^ 65)
    @printf("%-20s %-15s %-15s\n", "Method", "u(0)", "Abs. Error")
    println("-" ^ 65)
    @printf("%-20s %-15.6f %-15.6f\n", "Exact", u_exact(0), 0.0)
    @printf("%-20s %-15.6f %-15.6f\n", "Collocation (N=2)",
            u_approx(0, a0_coll, a1_coll),
            abs(u_exact(0) - u_approx(0, a0_coll, a1_coll)))
    @printf("%-20s %-15.6f %-15.6f\n", "Galerkin (N=2)",
            u_approx(0, a0_gal, a1_gal),
            abs(u_exact(0) - u_approx(0, a0_gal, a1_gal)))
    println("=" ^ 65)

    # RMS errors
    rms_coll = sqrt(sum(error_coll .^ 2) / N_POINTS)
    rms_gal  = sqrt(sum(error_gal .^ 2) / N_POINTS)
    @printf("\nRMS Error (Collocation): %.6f\n", rms_coll)
    @printf("RMS Error (Galerkin):    %.6f\n", rms_gal)

    # Max errors
    max_coll = maximum(abs.(error_coll))
    max_gal  = maximum(abs.(error_gal))
    @printf("\nMax Error (Collocation): %.6f\n", max_coll)
    @printf("Max Error (Galerkin):    %.6f\n", max_gal)

    return fig
end

# Run if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
